compiler/ras/Debug.cpp

```diff
@@ -727,7 +727,8 @@ TR_Debug::printInstrDumpHeader(const char * title)
 uint8_t *
 TR_Debug::printPrefix(TR::FILE *pOutFile, TR::Instruction *instr, uint8_t *cursor, uint8_t size)
    {
-   if (cursor != NULL)
+   // Code start is unknown until after the binary encoding phase
+   if (_comp->cg()->getCodeGeneratorPhase() >= TR::CodeGenPhase::BinaryEncodingPhase && cursor >= _comp->cg()->getCodeStart())
```

Comment:
Could also do:

```
if (_comp->cg()->getCodeStart() != NULL && cursor >= _comp->cg()->getCodeStart())
```

Not sure which is better. @aviansie-ben what are your thoughts on this?

Reply:
Neither of these solutions would really fix the underlying problem... Performing a comparison between pointers using relational operators is undefined behaviour if they're not from the same array or struct [1], so we can't use that to check for invalid pointers here. It's also very possible that we're invoking UB in generating these invalid pointers in the first place. The best way to tackle this IMO is to prevent invalid pointers like that from getting here in the first place, since we're already playing with fire if we let them get here.

[1] https://en.cppreference.com/w/c/language/operator_comparison#Relational_operators

Reply:
I suspect that would involve looking at all places that reach this function either directly or indirectly and checking the cursor is valid prior to calling the `printPrefix` function. That list is rather large (every single snippet on all platforms). Do you think that is feasible?

Alternatively we can just prohibit printing of snippets which have not been binary encoded.

Reply:
Hmm... The number of places that would need to be changed does indeed look to be pretty large, since logic is duplicated between basically every snippet. However, it doesn't look like just fixing `printPrefix` would suffice, even disregarding the UB. After a very brief search, I've already found a number of places across the various codegens that will dereference these invalid pointers without going through `printPrefix`:

https://github.com/eclipse/omr/blob/0e780bb7e7d2c3c449ca8f2b2500b6121b8be00e/compiler/z/codegen/ConstantDataSnippet.cpp#L651
https://github.com/eclipse/omr/blob/0e780bb7e7d2c3c449ca8f2b2500b6121b8be00e/compiler/x/codegen/DataSnippet.cpp#L178
https://github.com/eclipse/omr/blob/0e780bb7e7d2c3c449ca8f2b2500b6121b8be00e/compiler/arm/codegen/ARMDebug.cpp#L839

There are likely many more that I've missed that dereference these pointers indirectly, but overall it looks like there is no easy, centralized fix that would allow snippets to be printed before binary encoding. Unfortunately, I think preventing snippets from printing before binary encoding is complete might be the only way to go unless somebody's willing to fix every individual snippet.

Reply:
Yikes! You're totally correct. I don't see an easy way of fixing this either. I think the safest thing may be to prevent printing of snippets before binary encoding as the code was clearly not designed to handle this case. I'll open up an issue so we don't forget to circle back on this if/when we get to refactoring the pretty printing of codegen objects.

Reply:
Opened up #5111 to address this. I'm going to close this PR as it does not fully address the issue at hand.